docs/source/contributing.rst

```diff
@@ -179,6 +188,40 @@ After running the tests, you can inspect the test coverage by opening
 in coverage is also reported by the codecov bot. Ideally, the test coverage
 should go up with each pull request, at least it should stay the same.
 
+.. _`benchmarking`:
+
+Benchmarking
+~~~~~~~~~~~~
+
+LiberTEM uses `pytest-benchmark
+<https://pytest-benchmark.readthedocs.io/en/latest/usage.html>`_ to benchmark
+certain performance-critical parts of the code. Benchmarking is disabled by
+default. You can enable the benchmarks ad-hoc using
+
+.. code-block:: shell
+
+   $ pytest --benchmark-enable benchmarks/
```

Comment:
This may need to be updated with good parameters for interactive use, as opposed to the continuous approach that will extract good data from whatever was recorded. For example, we may want to enable warmup here, and possibly set a minimum number of rounds. Here are a few runs of the same functions (`dd_1` has the same function body as `dd_2`, but is a separate numba-jitted function) using the same settings:

```
-------------------------------------------------------------------------------------- benchmark: 2 tests -------------------------------------------------------------------------------------
Name (time in us)                Min                Max              Mean            StdDev            Median               IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_default_decode[dd_2]     1.5410 (1.0)      14.9130 (1.86)     2.3738 (1.0)      1.1634 (1.0)      2.5880 (1.50)     1.3660 (4.34)          2;2      421.2574 (1.0)         241           1
test_default_decode[dd_1]     1.6180 (1.05)      7.9990 (1.0)      2.7780 (1.17)     2.5606 (2.20)     1.7280 (1.0)      0.3150 (1.0)           1;1      359.9700 (0.85)          6           1
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

------------------------------------------------------------------------------------- benchmark: 2 tests -------------------------------------------------------------------------------------
Name (time in us)                Min               Max              Mean            StdDev            Median               IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_default_decode[dd_2]     1.5300 (1.0)      7.2120 (1.01)     1.5911 (1.0)      0.3562 (1.0)      1.5615 (1.0)      0.0230 (1.0)          2;11      628.4784 (1.0)         254           1
test_default_decode[dd_1]     1.6580 (1.08)     7.1320 (1.0)      2.6458 (1.66)     2.2017 (6.18)     1.7095 (1.09)     0.3460 (15.05)         1;1      377.9547 (0.60)          6           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

```

At least, we should set `--benchmark-warmup=true`, here are two results with that enabled:


```
-------------------------------------------------------------------------------------- benchmark: 2 tests --------------------------------------------------------------------------------------
Name (time in us)                Min               Max              Mean            StdDev            Median               IQR              Outliers  OPS (Kops/s)            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_default_decode[dd_1]     1.0856 (1.0)      5.9791 (1.37)     1.2634 (1.0)      0.3552 (1.12)     1.1335 (1.0)      0.0113 (1.0)     11044;16941      791.5423 (1.0)       91134          10
test_default_decode[dd_2]     1.1231 (1.03)     4.3800 (1.0)      1.2819 (1.01)     0.3178 (1.0)      1.1723 (1.03)     0.0164 (1.45)     9828;14143      780.0650 (0.99)      88301          10
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

--------------------------------------------------------------------------------------- benchmark: 2 tests --------------------------------------------------------------------------------------
Name (time in us)                Min               Max              Mean            StdDev            Median               IQR               Outliers  OPS (Kops/s)            Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_default_decode[dd_2]     1.0436 (1.0)      4.8358 (1.0)      1.2549 (1.01)     0.3973 (1.13)     1.0919 (1.0)      0.0142 (1.14)     13742;20244      796.8563 (0.99)      94581          10
test_default_decode[dd_1]     1.0656 (1.02)     4.8628 (1.01)     1.2448 (1.0)      0.3508 (1.0)      1.1117 (1.02)     0.0125 (1.0)      12109;17419      803.3550 (1.0)       93258          10
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

```

Maybe we can explicitly disable warmup just in tox, to make "interactive" use less foot-gunn-y? And put in a note that the tox run will not have warmup rounds...

A question to look at then is if the run without warmup will have enough statistics to be useful. Maybe we **can** do a warmup run, but include those numbers in the output data somehow? Might need changes in pytest-benchmarks...